Mono require the fallback for Mac (and probably Linux) distributions

simonferquel opened the pull request:
After some tests on mac, it seems the Mac version of mono has the "_headers" field... so we need the fallback for mono as well :/.

simonferquel committed:
Mono require the fallback for Mac (and probably Linux) distributions

Signed-off-by: Simon Ferquel <simon.ferquel@unity3d.com>

TalAloni commented on StandardSocketsHttpHandler/Net/Http/Extensions/HttpRequestMessageExtensions.cs:
```diff
@@ -13,9 +13,9 @@ public static bool HasHeaders(this HttpRequestMessage request)
             bool isDotNetFrameworkOrMono = isDotNetFramework || RuntimeUtils.IsMono();
```
So we don't need the 'isDotNetFramework' variable anymore...

simonferquel replied:
Indeed. And to avoid repetitive costly reflection calls, I'll cache the fieldinfo into a static field.

simonferquel replied:
And I will drop the framework dependant fallback: I'll just try with "headers" then "_headers"

simonferquel commented:
@TalAloni wdyt of this change ? I resolve the field info once and for all, and in the same manner for all runtimes : check for either _headers or headers field.

TalAloni commented:
You went a bit too far I'm afraid - I'm all for optimizations but you need to decide which goal you want to promote in this PR - better performance or Mono on Mac support - each should have its own PR.

simonferquel commented:
Well it is a code simplification that makes it work for Mono on mac (and any runtime which would decide to change its field naming convention between 2 versions), which as a side effect is also an optimization :).

TalAloni commented on StandardSocketsHttpHandler/Net/Http/Extensions/HttpRequestMessageExtensions.cs:
```diff
@@ -6,20 +6,14 @@ namespace System.Net.Http
     {
+        // depending on runtime the backing field for headers is either _headers or headers
+        // _headers in .net core and some .net fx and mono versions.
+        // headers in other .net fx / mono versions.
```
Please be more descriptive and detail the info we already know ('.NET Framework 4.6.1', 'Mono on Mac')

TalAloni commented on StandardSocketsHttpHandler/Net/Http/Extensions/HttpRequestMessageExtensions.cs:
```diff
@@ -6,20 +6,14 @@ namespace System.Net.Http
 {
     internal static class HttpRequestMessageExtensions
     {
+        // depending on runtime the backing field for headers is either _headers or headers
```
Please start with a capital letter ("Depending")

TalAloni commented on StandardSocketsHttpHandler/Net/Http/Extensions/HttpRequestMessageExtensions.cs:
```diff
@@ -6,20 +6,14 @@ namespace System.Net.Http
+        // depending on runtime the backing field for headers is either _headers or headers
+        // _headers in .net core and some .net fx and mono versions.
+        // headers in other .net fx / mono versions.
+        static FieldInfo _headersField = typeof(HttpRequestMessage).GetField("_headers", BindingFlags.Instance | BindingFlags.NonPublic) ??
```
please name the variable 's_headersField'

TalAloni replied:
I prefer 'private static'

simonferquel committed:
More details in comments

Signed-off-by: Simon Ferquel <simon.ferquel@unity3d.com>

simonferquel committed:
Code style adjustments

Signed-off-by: Simon Ferquel <simon.ferquel@unity3d.com>

TalAloni commented on StandardSocketsHttpHandler/Net/Http/Extensions/HttpRequestMessageExtensions.cs:
```diff
@@ -6,20 +6,14 @@ namespace System.Net.Http
+        // headers in other .net fx / mono versions.
+        static FieldInfo _headersField = typeof(HttpRequestMessage).GetField("_headers", BindingFlags.Instance | BindingFlags.NonPublic) ??
+            typeof(HttpRequestMessage).GetField("headers", BindingFlags.Instance | BindingFlags.NonPublic);
         public static bool HasHeaders(this HttpRequestMessage request)
```
please add an empty line above

simonferquel committed:
Make s_headersField readonly

Signed-off-by: Simon Ferquel <simon.ferquel@unity3d.com>

simonferquel commented:
- Added more details in comment about versions we know of that have the "_headers" field name
- Adjusted code style according to review comments
- Marked the static field readonly

TalAloni commented on StandardSocketsHttpHandler/Net/Http/Extensions/HttpRequestMessageExtensions.cs:
```diff
@@ -6,20 +6,15 @@ namespace System.Net.Http
         {
-            // Note: The field name is _headers in .NET core 
-            bool isDotNetFramework = RuntimeUtils.IsDotNetFramework();
-            bool isDotNetFrameworkOrMono = isDotNetFramework || RuntimeUtils.IsMono();
```
So we can remove the IsMono method as well

simonferquel replied:
We can remove the whole RuntimeInformation class that is not used anymore. Should I do it in this PR?

simonferquel replied:
RuntimeUtils I mean

TalAloni replied:
Please remove just the IsMono method

simonferquel committed:
Remove RuntimeUtils.IsMono()

Signed-off-by: Simon Ferquel <simon.ferquel@unity3d.com>

TalAloni commented:
Thanks,
The change is included in 3.1.0.4
https://www.nuget.org/packages/StandardSocketsHttpHandler/3.1.0.4
